Let Express set port for test environment.

hiyaryan opened the pull request:
Removes MEMORY environment variable.
Uses yargs to pass arguments to the application on launch.
 - Adds yargs to dependecies list.
 - Passes port=3001 to test instances of the app. Resolves lint error in index.test.ts regarding imports out of order.

hiyaryan commented:
Note that these changes allows the development container to be up. In other words, the development container may be running simultaneously with the test environment because it runs on a different port.

davidzlu reviewed (changes requested):
I'm totally cool with removing the MEMORY env variable. I am getting a weird error trying to run the tests locally though. I suggested a different way of handling the ports. It's kinda dirty, but it seems to work too.

hiyaryan commented:
@davidzlu Per your review I have reverted to pre-`yargs` code and wrote a new utility function to dynamically assign a port based on the `PORT` in `.env` called `getPort`. It does this by incrementing the port by 1 with checks to ensure that the port is available to use using a `isPortInUse` in the same util file. With that I updated `index.ts` and it's associated test file to use the `getPort` function. And I added unit tests for `getPort`. 

What's good about having this function is that we no longer require additional dependencies, it feels like a better solution than using hardcoded port values, and we can also further enhance the code for convenience to check that the `PORT` in .env is available in the development environment.

davidzlu commented on backend/src/utils/getPort.ts:
```diff
@@ -0,0 +1,54 @@
+ * 
+ * @returns A promise that resolves to the port number.
+ */
+export default async function getPort(): Promise<number> {
```
This is an interesting way of doing this. I think it can work, but I'm also wondering why you decided to go with this over letting express choose a random unused port by passing in 0 for the port number (https://expressjs.com/en/4x/api.html#app.listen)?

I asked an LLM how to get a port number that way and this is what it gave me:
```
const server = app.listen(() => {
  const address = server.address();
  if (address && typeof address !== 'string') {
    const port = address.port;
    console.log(`Server is running on port ${port}`);
  }
});
```

hiyaryan replied:
I have no strong reason to do it this way, I think I just got carried away. 😁 Thanks for pointing out that we can just let express do it!

davidzlu commented on backend/src/utils/getPort.ts:
```diff
@@ -0,0 +1,54 @@
+  else port = parseInt(port, 10);
+  
+  // Increment the port by 1 if the environment is test.
+  if (process.env.NODE_ENV === 'test') port += 1;
```
So here you could return 0 and then dynamically look up what port number got assigned in the test environment instead of testing out ports incrementally

davidzlu commented on backend/tests/index.test.ts:
```diff
@@ -14,13 +14,16 @@ afterAll(async () => {
 describe('Server startup', () => {
-  it('should start the server in non-production mode', (done) => {
+  it('should start the server in non-production mode', async () => {
     process.env.NODE_ENV = 'development';
```
If you leave this line in then the test would use the port defined in .env right? Is that the intent?

davidzlu commented on backend/tests/utils/getPort.test.ts:
```diff
@@ -0,0 +1,53 @@
+    expect(port).toBe(3000);
+  });
+
+  it('should increment the port by 1 if NODE_ENV is "test"', async () => {
```
I think there's an extremely small chance that this test is non-deterministic because the test might run in a host that is using the 3001 port already

davidzlu reviewed (changes requested):
Great effort with the port stuff! I like that it's taking care of all the environments, but I think it might make more sense to let express handle finding a port rather than searching for one with a utility function, unless you've got a strong reason to not use it.

hiyaryan commented:
Great update @davidzlu! 🙌 An elegant and simple solution. I pushed a followup commit that just completely removes the `getPort` utility function (-1 for the `utils` directory). Somewhat of a revert with a little refactoring and code clean up.

davidzlu reviewed (approved):
Everything seems good! Great work!
